services/vod-to-live/service/src/domains/validation/setup-pre-publishing-validation-webhook.ts

```diff
@@ -0,0 +1,87 @@
+import {
+  Dict,
+  generateWebhookResponse,
+  handleWebhookErrorMiddleware,
+  Logger,
+  verifyWebhookRequestMiddleware,
+  WebhookRequestMessage,
+  WebhookResponse,
+} from '@axinom/mosaic-service-common';
+import { Express, Request, Response } from 'express';
+import { Config, ValidationErrors } from '../../common';
+import { AzureStorage } from '../azure';
+import { ChannelPublishedValidationWebhookHandler } from './channel-published-validation-handler';
+import { PlaylistPublishedValidationWebhookHandler } from './playlist-published-validation-handler';
+import {
+  PrePublishingPayload,
+  ValidationWebhookHandler,
+} from './validation-webhook-model';
+
+const logger = new Logger({
+  context: 'Pre-Publishing-Validation-Webhook-Handler',
+});
+
+export const setupPrePublishingValidationWebhook = (
+  app: Express,
+  config: Config,
+  storage: AzureStorage,
+): void => {
+  const handlers: ValidationWebhookHandler<PrePublishingPayload>[] = [
+    new ChannelPublishedValidationWebhookHandler(config),
+    new PlaylistPublishedValidationWebhookHandler(config, storage),
+  ];
+  app.post(
+    '/pre-publishing',
+    verifyWebhookRequestMiddleware({
+      webhookSecret: config.prePublishingWebhookSecret,
+    }),
+    handleWebhookErrorMiddleware,
+    async (
+      req: Request<
+        Dict<string>,
+        Dict<string>,
+        WebhookRequestMessage<PrePublishingPayload>
+      >,
+      res: Response,
+    ) => {
+      const requestBody = req.body;
+      let validationResponse: WebhookResponse<PrePublishingPayload> | null =
+        null;
+      const handler = handlers.find((h) => h.canHandle(requestBody));
+      if (handler) {
+        try {
+          const result = await handler.handle(requestBody);
+          validationResponse =
+            generateWebhookResponse<PrePublishingPayload>(result);
+        } catch (error) {
+          logger.error({
+            message: `The message of type ${requestBody.message_type} could not be validated.`,
+            details: {
+              requestBody: requestBody,
+              error: error,
+            },
+          });
+          validationResponse = generateWebhookResponse<PrePublishingPayload>({
+            payload: null,
+            errors: [ValidationErrors.ValidationFailed],
+            warnings: [],
+          });
+        }
+      } else {
+        logger.warn({
+          message: `The received message cannot be handled. Its type is: ${requestBody.message_type}.`,
+          details: { ...requestBody },
+        });
+        // generating response to the caller
+        validationResponse = generateWebhookResponse<PrePublishingPayload>({
+          payload: null,
+          errors: [ValidationErrors.ValidationNotPossible],
+          warnings: [],
+        });
+      }
+      //returns status code 400, if validation result has any errors, otherwise return 200
+      const statusCode = validationResponse.errors.length > 0 ? 400 : 200;
```

Comment:
Up for discussion: The way I see it, if an error is thrown for whatever reason - we need to generate a fitting response and send it with code 400+ (I guess 400 Bad Request is ok).

On the other hand, if no explicit error was thrown, and the validation handler returned some errors inside of the payload - we should consider it as a valid response (with errors) and return 200? right now it would respond with 400, which I'm not certan is correct...

Reply:
What is our general pattern here? I suggest returning a 400 error code in case of errors. So we should IMO instead adjust the other places. For GraphQL this is of course another story - there the specification dictates a 200 response with an error object. But for REST-based webhooks, we should consider the HTTP codes IMO.

Reply:
We don't have enough webhooks to have a general pattern yet (and the entitlement ones are not producing custom validation messages). I was just debating if validation error is equal to thrown error. As long as returned payload errors are correctly visualized and processed - I guess it should be fine.